libs-scala/scala-utils/src/main/scala/com/daml/scalautil/nonempty/NonEmpty.scala

```diff
@@ -116,20 +145,28 @@ object NonEmptyColl extends NonEmptyCollInstances {
     def ++(that: IterableOnce[A]): NonEmpty[C] = un((self: ESelf) ++ that)
   }
 
-  implicit final class NEPreservingOps[A, C](
-      private val self: NonEmpty[IterableOps[A, imm.Iterable, C with imm.Iterable[A]]]
+  implicit final class NEPreservingOps[A, CC[X] <: imm.Iterable[X], C](
+      private val self: NonEmpty[IterableOps[A, CC, C with imm.Iterable[A]]]
   ) extends AnyVal {
     import NonEmpty.{unsafeNarrow => un}
-    private type ESelf = IterableOps[A, imm.Iterable, C with imm.Iterable[A]]
+    private type ESelf = IterableOps[A, CC, C with imm.Iterable[A]]
+    def groupBy[K](f: A => K): NonEmpty[Map[K, NonEmpty[C]]] =
+      NonEmpty.subst[Lambda[f[_] => f[Map[K, f[C]]]]]((self: ESelf) groupBy f)
+    def groupBy1[K](f: A => K): NonEmpty[Map[K, NonEmpty[C]]] = self groupBy f
     def toList: NonEmpty[List[A]] = un((self: ESelf).toList)
     def toVector: NonEmpty[Vector[A]] = un((self: ESelf).toVector)
     def toSeq: NonEmpty[imm.Seq[A]] = un((self: ESelf).toSeq)
     def toSet: NonEmpty[Set[A]] = un((self: ESelf).toSet)
     def toMap[K, V](implicit isPair: A <:< (K, V)): NonEmpty[Map[K, V]] = un((self: ESelf).toMap)
-    def to[C1 <: imm.Iterable[A]](factory: Factory[A, C1]) = un((self: ESelf) to factory)
+    def to[C1 <: imm.Iterable[A]](factory: Factory[A, C1]): NonEmpty[C1] = un(
+      (self: ESelf) to factory
+    )
+    def zipWithIndex: NonEmpty[CC[(A, Int)]] = un((self: ESelf).zipWithIndex)
```

Comment:
Nit: If you were adding the type annotation above to be consistent, perhaps add it here too?